display_list/display_list_unittests.cc

```diff
@@ -3781,5 +3800,87 @@ TEST_F(DisplayListTest, SaveLayerBoundsClipDetectionSimpleClippedRect) {
   EXPECT_TRUE(expector.all_bounds_checked());
 }
 
+class DepthExpector : public virtual DlOpReceiver,
+                      virtual IgnoreAttributeDispatchHelper,
+                      virtual IgnoreTransformDispatchHelper,
+                      virtual IgnoreClipDispatchHelper,
+                      virtual IgnoreDrawDispatchHelper {
+ public:
+  explicit DepthExpector(std::vector<uint32_t> expectations)
+      : depth_expectations_(std::move(expectations)) {}
+
+  void save() override {
+    // This method should not be called since we override the variant with
+    // the total_content_depth parameter.
+    FAIL() << "save(no depth parameter) method should not be called";
+  }
+
+  void save(uint32_t total_content_depth) override {
+    ASSERT_LT(index_, depth_expectations_.size());
+    EXPECT_EQ(depth_expectations_[index_], total_content_depth)
+        << "at index " << index_;
+    index_++;
+  }
+
+  void saveLayer(const SkRect& bounds,
+                 SaveLayerOptions options,
+                 const DlImageFilter* backdrop) override {
+    // This method should not be called since we override the variant with
+    // the total_content_depth parameter.
+    FAIL() << "saveLayer(no depth parameter) method should not be called";
+  }
+
+  void saveLayer(const SkRect& bounds,
+                 const SaveLayerOptions& options,
+                 uint32_t total_content_depth,
+                 const DlImageFilter* backdrop) override {
+    ASSERT_LT(index_, depth_expectations_.size());
+    EXPECT_EQ(depth_expectations_[index_], total_content_depth)
+        << "at index " << index_;
+    index_++;
+  }
+
+ private:
+  size_t index_ = 0;
+  std::vector<uint32_t> depth_expectations_;
+};
+
+TEST_F(DisplayListTest, SaveContentDepthTest) {
+  DisplayListBuilder child_builder;
+  child_builder.DrawRect({10, 10, 20, 20}, DlPaint());  // depth 1
+  auto child = child_builder.Build();
+
+  DisplayListBuilder builder;
+  builder.DrawRect({10, 10, 20, 20}, DlPaint());  // depth 1
+
+  builder.Save();  // covers depth 1->9
+  {
+    builder.Translate(5, 5);
+    builder.DrawRect({10, 10, 20, 20}, DlPaint());  // depth 2
+
+    builder.DrawDisplayList(child, 1.0f);  // depth 3 (content) + 4 (self)
+
+    builder.SaveLayer(nullptr, nullptr);  // covers depth 5->6
+    {
+      builder.DrawRect({12, 12, 22, 22}, DlPaint());  // depth 5
+      builder.DrawRect({14, 14, 24, 24}, DlPaint());  // depth 6
+    }
+    builder.Restore();  // layer is restored with depth 7
+
+    builder.DrawRect({16, 16, 26, 26}, DlPaint());  // depth 8
+    builder.DrawRect({18, 18, 28, 28}, DlPaint());  // depth 9
+  }
+  builder.Restore();
+
+  builder.DrawRect({16, 16, 26, 26}, DlPaint());  // depth 10
+  builder.DrawRect({18, 18, 28, 28}, DlPaint());  // depth 11
```

Comment:
Could you add a test that checks the depth of clip ops added to both the root and to the root and within a SaveLayer? That way we know they close with the correct depth (max depth of everything they affect).

Reply:
What is the "depth of clip ops"? They aren't accounted in DL. That's an Impeller thing...?

All DL does is count how many render calls are within a save/restore or an entire DL. Clips are invisible to that.

Reply:
As in the class comment in dl_op_receiver, the only places depth information is reported is the DL total_depth and the argument passed to save/saveLayer.